### PR TITLE
Serialization of Oscar data using Polymake

### PR DESCRIPTION
**Pull request by alexej-jordan**
For now, this branch contains a raw working version for saving/loading rational ideals. Saving is accomplished by converting the relevant data to types polymake knows. An "empty" `Polymake.BigObject` is created (an `Ideal` in this case, but that is not relevant for the process) and receives the information as attachments. This object can then be saved using polymake's serialization. Loading works exactly the other way around.

I am not sure how I should handle compatibilty, though. On one point I enforce `keep_ordering = false` for `BiPolyArray`s, which might not be correct for every case. In a former version this lead to the inability to compare the initial ideal with the saved/loaded one due to the Singular ring having a different order even though the other values look correct.

### CODE REVIEW

File: src/Rings/mpoly-ideals.jl [hunk 1]
```diff
@@ -46,15 +47,15 @@ function ideal(R::MPolyRing, g::Vector)
   return ideal(f)
 end
 
-function ideal(Qxy::MPolyRing{T}, x::MPolyElem{T}) where T <: RingElem 
+function ideal(Qxy::MPolyRing{T}, x::MPolyElem{T}) where T <: RingElem
```
Comment (fingolfin) [inline src/Rings/mpoly-ideals.jl]: Can the white space changes please either be removed or be put into a commit of their own?

**Comment by fingolfin** [comment]
I am a bit confused as what the eventual goal here is? Is this to scratch a particular itch (you have a project where you need to (de)serialize a bunch of ideals)? Or is this intended as a start for a more extensive (de)serialization system for OSCAR? This then would raise the question: Do we want to store data in a way so that other systems might be able to read them at some point in the future? Also, what is the motivation for serializing Oscar ideals through Polymake machinery, instead of e.g. https://github.com/JuliaIO/JLD.jl  or https://github.com/JuliaIO/JLD2.jl ? (Honest, open question, i.e., I am not saying we shouldn't, it just would be good if this was spelled out).

**Comment by micjoswig** [comment]
> I am a bit confused as what the eventual goal here is? Is this to scratch a particular itch (you have a project where you need to (de)serialize a bunch of ideals)? Or is this intended as a start for a more extensive (de)serialization system for OSCAR? This then would raise the question: Do we want to store data in a way so that other systems might be able to read them at some point in the future? Also, what is the motivation for serializing Oscar ideals through Polymake machinery, instead of e.g. https://github.com/JuliaIO/JLD.jl or https://github.com/JuliaIO/JLD2.jl ? (Honest, open question, i.e., I am not saying we shouldn't, it just would be good if this was spelled out).

We had talked about loading/storing Oscar data in the context of MaRDI.  I think this is very important.  The current draft should be seen as a proof of concept. Using the polymake machinery is viable, as Alex's code shows.  I am open to other suggestions, but then they should have the same (or more) features.

While JLD is interesting, too, it serves a different purpose.  I would be in favor to use JLD, but only in addition to another mechanism.

Let me explain (only briefly).  The polymake file format provides a [general approach for abstraction of mathematical data](http://link.springer.com/chapter/10.1007/978-3-319-42432-3_50).  Among its most important features: versioning and data migration; data base ready (using MongoDB).

In a nutshell: This file format is designed to store data in a way such that you should be able to load it in some Oscar version of the year 2041.  At least this is what we can do looking backwards: Oscar 0.6.0 can load polymake polytope files from 1997, and this mechanism is transparent for the user.

JLD can only load data that was stored in a julia version which stores those data in a binary compatible way, which is beyond our control and most likely subject to changes.  But as I said: using JLD in addition  would be nice, too.

**Comment by thofma** [comment]
This needs much more thought because of the polynomial ring or the parent/element system in general. The tests here are working by accident. If the original parent had not been cached (`chache = false`), the loaded ideal will not be equal to the original one I think.

**Comment by micjoswig** [comment]
> This needs much more thought because of the polynomial ring or the parent/element system in general. The tests here are working by accident. If the original parent had not been cached (`chache = false`), the loaded ideal will not be equal to the original one I think.

Sure.  That's why this is currently a proof of concept only.

**Comment by fieker** [comment]
On Wed, Oct 13, 2021 at 03:58:41AM -0700, Tommy Hofmann wrote:
> This needs much more thought because of the polynomial ring or the parent/element system in general. The tests here are working by accident. If the original parent had not been cached (`chache = false`), the loaded ideal will not be equal to the original one I think. 
> 
Possibly every generator in a different ring...

**Comment by micjoswig** [comment]
> On Wed, Oct 13, 2021 at 03:58:41AM -0700, Tommy Hofmann wrote: This needs much more thought because of the polynomial ring or the parent/element system in general. The tests here are working by accident. If the original parent had not been cached (`chache = false`), the loaded ideal will not be equal to the original one I think.
> Possibly every generator in a different ring...
> […](#)
> -- You are receiving this because you are subscribed to this thread. Reply to this email directly or view it on GitHub: [#731 (comment)](https://github.com/oscar-system/Oscar.jl/pull/731#issuecomment-942179704)

This can easily be resolved by specifying the ring information (in whichever form you like), e.g., as an attachment.  You just need to say what you want.

File: src/Rings/mpoly-ideals.jl [hunk 2]
```diff
@@ -1127,3 +1128,72 @@ function isone(I::MPolyIdeal)
   gb = groebner_basis(I, complete_reduction = true)
   return isconstant(gb[1]) && isunit(first(coefficients(gb[1])))
 end
+
+function Base.convert(::Type{Polymake.Polynomial{Polymake.Rational, Polymake.to_cxx_type(Int64)}}, x::fmpq_mpoly)
+  c::Vector{Polymake.Rational} = collect(coefficients(x))
+  e::Matrix{Int64} = vcat(transpose.(exponent_vectors(x))...)
```
Comment (tthsqe12) [inline src/Rings/mpoly-ideals.jl]: Minor point. I think this kind of splatting can be potentially problematic because it will compile code for every polynomial length of x that has been passed. Something like this will do something similar and be compiled only once.
```
julia> R, (x, y) = PolynomialRing(QQ, ["x", "y"]);

julia> a = (1+x+y)^3
x^3 + 3*x^2*y + 3*x^2 + 3*x*y^2 + 6*x*y + 3*x + y^3 + 3*y^2 + 3*y + 1

julia> reduce(hcat, exponent_vectors(a))
2×10 Matrix{Int64}:
 3  2  2  1  1  1  0  0  0  0
 0  1  0  2  1  0  3  2  1  0

julia> reduce(vcat, transpose.(exponent_vectors(a)))
10×2 Matrix{Int64}:
 3  0
 2  1
 2  0
 1  2
 1  1
 1  0
 0  3
 0  2
 0  1
 0  0
```

Comment (tthsqe12) [reply src/Rings/mpoly-ideals.jl]: Also, the existing code does not do well when there are no exponent vectors, i.e. x is zero.

**Comment by alexej-jordan** [comment]
> This needs much more thought because of the polynomial ring or the parent/element system in general. The tests here are working by accident. If the original parent had not been cached (`chache = false`), the loaded ideal will not be equal to the original one I think.

You are right, some solid decisions need to be made. The tests i included right now are meant to demonstrate functionality and syntax.

For the parent/element system i thought about having an optional parent argument for the `load_ideal` function. And, because i do not know all the kinks of that system yet, i would happily hear suggestions or encourage a discussion on this. Independently of the implementation, this is something that needs to be cleared sooner or later if we want to save and load ideals (and other objects that come with a similar system).

**Comment by alexej-jordan** [comment]
> > On Wed, Oct 13, 2021 at 03:58:41AM -0700, Tommy Hofmann wrote: This needs much more thought because of the polynomial ring or the parent/element system in general. The tests here are working by accident. If the original parent had not been cached (`chache = false`), the loaded ideal will not be equal to the original one I think.
> > Possibly every generator in a different ring...
> > […](#)
> > -- You are receiving this because you are subscribed to this thread. Reply to this email directly or view it on GitHub: [#731 (comment)](https://github.com/oscar-system/Oscar.jl/pull/731#issuecomment-942179704)
> 
> This can easily be resolved by specifying the ring information (in whichever form you like), e.g., as an attachment. You just need to say what you want.

For the Oscar side of ring objects, the relevant information (including variable names and ordering; excluding field, because we only consider rationals for now, and cached) is attached to the `BigObject` while saving, and used for creating the ring while loading the object. Correct me if I am missing anything. As outlined in my previous comment, managing parents and elements is not trivial and might need some discussion before.

File: src/Rings/mpoly-ideals.jl (hunk 2 above)
Comment (alexej-jordan) [reply src/Rings/mpoly-ideals.jl]: Both fixed, thank you!

**Comment by micjoswig** [comment]
The problem is that, technically speaking, a ring in Oscar/Singular is a pointer. Right? This means that, in a way, any particular ring cannot be saved (in a meaningful way).  Period.

Here is a suggestion.

To recover the data it suffices to create a ring which is isomorphic to the original one; i.e., we need to know the coefficient type and the number of variables (and the names of the variables, if you want to keep them, too).  So a subsequent save/load combination creates a proper copy.  It does not recover the original.

The load (of an ideal) creates a new ring (from the coefficient type and the number of variables) and makes each polynomial among the given generators of the ideal (or any other polynomials to be associated, e.g., generators of some Gröbner basis) an element of that ring.

**Comment by fieker** [comment]
On Thu, Oct 14, 2021 at 12:33:51AM -0700, Michael Joswig wrote:
> The problem is that, technically speaking, a ring in Oscar/Singular is
> a pointer. Right? This means that, in a way, any particular ring
> cannot be saved (in a meaningful way).  Period.
> 
> Here is a suggestion.
> 
> To recover the data it suffices to create a ring which is isomorphic
> to the original one; i.e., we need to know the coefficient type and
> the number of variables (and the names of the variables, if you want
> to keep them, too).  So a subsequent save/load combination creates a
> proper copy.  It does not recover the original.
> 
> The load (of an ideal) creates a new ring (from the coefficient type
> and the number of variables) and makes each polynomial among the given
> generators of the ideal (or any other polynomials to be associated,
> e.g., generators of some Gröbner basis) an element of that ring.
>  
I think what we need are
 - a link to the specification of the serialisation - I don't think
   converting everything (especially those that do not exist there) to
   polymake is viable long term
 - define a framework to serialise arbitrary collections, or groups of
   objects. Within such a group, all elem-parent relations need to be
   maintained.
   My guess is that the interface will look quite similar to deepcopy:
   there needs to be a state objetc (e.g. IdDict) taking care of memorizing
   objects that were already serialised to make sure that only references
   are then send.
 - explicitly adding a parent to the deserialisation is only a stop-gap, in
   general there might be multiple parents involved for a range of
   sub-objects...
 - ideally I can use the same source code to also get some efficient,
   not-stable binary serialisation, maybe seperated by some IOContext?

serialise(::Context, ::Objetc, ::IdDict)?

**Comment by fingolfin** [comment]
What's the plan for this PR, @alexej-jordan ? Would you like to keep it open for a future update? Or can it be closed?
